File: src/main/kotlin/day2/day2.kt
package day2

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.nio.file.Paths

object Moves{
    const val ROCK = 1
    const val PAPER = 2
    const val SCISSORS = 3
}

object Result{
    const val WIN = 1
    const val DRAW = 2
    const val LOSE = 3

    fun getPoints(result: Int) : Int{
        return when(result){
            WIN -> 6
            DRAW -> 3
            LOSE -> 0
            else -> 0
        }
    }
}

fun main(){
    val rootDir = Paths.get(".").normalize().toAbsolutePath()
    var file : File = File("$rootDir/src/main/kotlin/day2/input.txt")
    var reader : BufferedReader = BufferedReader(FileReader(file))

    var str : String?

    var totalPoints = 0
    var totalPointsSecond = 0

    while(true){
        var roundPoints = 0

        str = reader.readLine()
        if(str == null){
            break
        }
        val splitStr = str.split(" ")

        val elfMove = getMove(splitStr[0])
        val myMove = getMove(splitStr[1])
        val result = getResult(splitStr[1])

        roundPoints += myMove
        println("You got $roundPoints for playing ${splitStr[1]}")

        roundPoints += getWinnerPoints(elfMove, myMove)
        totalPoints += roundPoints

        var roundPointsSecond = 0
        roundPointsSecond += getNeedMove(result, elfMove)
        roundPointsSecond += Result.getPoints(result)

        totalPointsSecond += roundPointsSecond
    }
    // First puzzle
    println("Total points: $totalPoints")

    // Second puzzle
    println("Total points: $totalPointsSecond")
}

fun getWinnerPoints(shapeElf: Int, myShape: Int): Int{
    if(shapeElf == myShape){
        return 3
    }
    if(shapeElf == Moves.PAPER && myShape == Moves.ROCK ||
            shapeElf == Moves.SCISSORS && myShape == Moves.PAPER ||
            shapeElf == Moves.ROCK && myShape == Moves.SCISSORS){
        return 0
    }
    return 6
}

fun getMove(shape: String) : Int {
    when(shape){
        in arrayOf("X", "A") -> return Moves.ROCK
        in arrayOf("Y", "B") -> return Moves.PAPER
        in arrayOf("Z", "C") -> return Moves.SCISSORS
    }
    return 0
}

fun getResult(result: String) : Int {
    when(result){
        "X" -> return Result.LOSE
        "Y" -> return Result.DRAW
        "Z" -> return Result.WIN
    }
    return 0
}

fun getNeedMove(result: Int, elfMove: Int): Int{
    if(result == Result.DRAW){
        return elfMove
    }
    if(result == Result.WIN){
        when(elfMove){
           Moves.ROCK -> return Moves.PAPER
           Moves.PAPER -> return Moves.SCISSORS
           Moves.SCISSORS -> return Moves.ROCK
        }
    }
    if(result == Result.LOSE){
        when(elfMove){
            Moves.ROCK -> return Moves.SCISSORS
            Moves.PAPER -> return Moves.ROCK
            Moves.SCISSORS -> return Moves.PAPER
        }
    }
    return 0
}

File: day2/day2.cpp
#include <iostream>
#include <fstream>
#include <string>

enum Result {
    WIN = 6,
    DRAW = 3,
    LOSE = 0,
};

enum Moves {
    ROCK = 1,
    PAPER = 2,
    SCISSOR = 3,
};

Moves playValue(char play);

Result getResult(char result);

Result gameResult(Moves elfMove, Moves myMove);

Moves neededMove(Moves elfMove, Result result);

int main(){
    std::ifstream file;
    file.open("input.txt");

    if(!file.is_open()){
        std::cout << "Error opening file!" << std::endl;
    }

    std::string line;

    int points = 0;

    int secPoints = 0;

    int roundPoints = 0;

    while(getline(file, line)){
        roundPoints = playValue(line[2]);
        roundPoints += gameResult(playValue(line[0]), playValue(line[2]));
        points += roundPoints;

        secPoints += neededMove(playValue(line[0]), getResult(line[2]));
        secPoints += getResult(line[2]);
    }

    std::cout << "Total points: " << points << std::endl;

    std::cout << "Total points: " << secPoints << std::endl;
    return 0;
}

Moves playValue(char play){
    switch(play){
        case 'A':
        case 'X': return ROCK;
        case 'B':
        case 'Y': return PAPER;
        case 'C':
        case 'Z': return SCISSOR;
    }
}

Result getResult(char result){
    switch(result){
        case 'X': return LOSE;
        case 'Y': return DRAW;
        case 'Z': return WIN;
    }
}

Result gameResult(Moves elfMove, Moves myMove){
    if(elfMove == myMove){
        return DRAW;
    }
    if(elfMove == ROCK && myMove == SCISSOR ||
        elfMove == PAPER && myMove == ROCK ||
        elfMove  == SCISSOR && myMove == PAPER){
            return LOSE;
        }
    return WIN;
}

Moves neededMove(Moves elfMove, Result result){
    if(result == DRAW){
        return elfMove;
    }
    if(result == LOSE){
        switch(elfMove){
            case ROCK: return SCISSOR;
            case PAPER: return ROCK;
            case SCISSOR: return PAPER;
        }
    }
    if(result == WIN){
        switch(elfMove){
            case ROCK: return PAPER;
            case PAPER: return SCISSOR;
            case SCISSOR: return ROCK;
        }
    }
}